### PR TITLE
57 & 58 Request/Push Handling Logic (cont)

### PR DESCRIPTION
**Pull request by chadlagore**
# What Changed

+ Implemented `Request` handling logic.
+ Finished `Push` handling logic.
    + This required an update to the `miner` module, @david-julien please 👀 .

### TODO 💡 

+ `Mine` should take an interface with a callback as an argument.
+ We can use this interface as a way to update progress on mining, alert the app of progress etc etc.

```go
type MinerUpdater interface {
    UpdateProgress(i int)
    UpdateFinished(result MiningResult)
}

go miner.Mine(bc, block, updater)

// As opposed to:
result := miner.Mine(bc, block)   // => Spins forever.
```

### CODE REVIEW

File: app/app.go
Comment (chadlagore) [inline app/app.go:205]: This is TODO.

File: app/app.go
Comment (bfbachmann) [inline app/app.go:98]: I'm doing some refactoring this week such that we don't have all these global variables, but instead we have instances of App for each Cumulus "node" we are running. The `chain` variable will also be part of that at some point just so you're aware.

File: app/app.go
Comment (bfbachmann) [inline app/app.go:130]: We should probably add a different error code here to be more specific since this technically isn't an invalid resource type error.

File: app/app.go
Comment (bfbachmann) [inline app/app.go]: I don't think you should check if the chain exists here because this function should only be called when we don't already have a blockchain right? Actually, either way, when a node boots it should always ask its peers for blocks anyway (regardless of whether it already has a blockchain stored locally) since there is no other way of knowing for sure that you have the most up-to-date blockchain.

File: app/app_test.go
Comment (bfbachmann) [inline app/app_test.go:17]: Why not `blockNumber` of type `int`?

File: miner/miner_test.go
Comment (david-julien) [inline miner/miner_test.go]: Reset the max target to its original value before the test returns so it doesn't cause any problems for any tests in the same file written after this test.

File: miner/miner.go
Comment (david-julien) [inline miner/miner.go]: Should CurrentlyMining and the lock be private?

**Review by david-julien — APPROVED** [review]

File: app/app.go
Comment (david-julien) [reply app/app.go:205]: Don't know how we're all handling TODOs but should we mark this as TODO in the comment just to make sure we don't forget?

File: miner/miner.go
Comment (chadlagore) [reply miner/miner.go]: Good call. Originally I wanted to have other packages check on them, but the exported function does that now.

File: blockchain/blockchain.go
Comment (bfbachmann) [inline blockchain/blockchain.go]: What if this is an out-of-bounds reference? I think you can also aggregate this line and the one below into one.

File: miner/miner.go
Comment (bfbachmann) [inline miner/miner.go:17]: I guess this might be overkill but you could put `currentlyMining` and `currentlyMiningLock` into a struct of type `Miner`. Then you could make the functions below receivers of the `Miner` type. That way we can manage multiple instances of `Miner`.

File: miner/miner.go
Comment (bfbachmann) [inline miner/miner.go:21]: You only need to set the first constant to `iota`, the compiler will fill in the following ones on its own (see https://golang.org/ref/spec#Constant_declarations and the section below).

File: miner/miner.go
Comment (bfbachmann) [inline miner/miner.go]: I know you added this earlier but we should probably add more detail here. Generally when we surface errors we want to know what package they're from and what action what being taken when the error occurred.
In this case something like "Mine failed, miner received an invalid block".

File: miner/miner.go
Comment (bfbachmann) [inline miner/miner.go]: Why have this be a type when we can just return `(complete bool, info int)` from `Mine()`? Do we use MiningResult outside this file?

File: miner/miner.go
Comment (bfbachmann) [inline miner/miner.go:85]: I love the idea of cancelling a mining job. Nice work, we'll definitely need to do this when we are notified of new blocks that were added to the chain that contain transactions we are trying to mine in this block.

**Review by bfbachmann — COMMENTED** [review]
Nice work

File: miner/miner.go
Comment (bfbachmann) [reply miner/miner.go:17]: Edit: I see you want to make `Mine()` take an interface and a callback as a means of updating progress. I think using a `Miner` type could help with that. The only thing is, I fell like in mining there is no notion of progress since you can never know for sure how close you are to the solution.

File: app/app.go
Comment (chadlagore) [reply app/app.go:98]: Good stuff :+1:

File: miner/miner.go
Comment (chadlagore) [reply miner/miner.go]: Actually, I don't even think the miner should be worried about validation. Good eye 👀

File: app/app_test.go
Comment (chadlagore) [reply app/app_test.go:17]: 😆

**Review by jordanschalm — APPROVED** [review]
🚢

File: miner/miner.go
Comment (chadlagore) [reply miner/miner.go:17]: 1. As for `Miner` type, I agree. Not in scope here, but I think we should do it. It would be nice to build in the ability to scale your miners outwards instead of upwards.
2. By progress I more meant, how many hashes it has attempted. Progress = `number_attempted / number_of_hashes_to_needed_to_likely_produce_a_match`.

**Comment by jordanschalm** [comment]
@chadlagore Feel free to merge this after review comments and merge conflicts are resolved 👍

**Comment by jordanschalm** [comment]
closes #56 
closes #57 
closes #58 

Let's close these and make more descriptive issues next meeting

**Pull request merged**
